src/adapter/src/rbac.rs

```diff
@@ -0,0 +1,236 @@
+            }
+        }
+    }
+}
+
+impl Error for UnauthorizedError {}
+
+pub fn check_command(catalog: &Catalog, cmd: &Command) -> Result<(), UnauthorizedError> {
+    if !catalog.system_config().enable_rbac_checks() {
+        return Ok(());
+    }
+
+    match cmd {
+        Command::DumpCatalog { session, .. } => {
+            if session.is_superuser() {
+                Ok(())
+            } else {
+                Err(UnauthorizedError::superuser("dump catalog".into()))
+            }
+        }
+        Command::Startup { .. }
+        | Command::Declare { .. }
+        | Command::Describe { .. }
+        | Command::VerifyPreparedStatement { .. }
+        | Command::Execute { .. }
+        | Command::StartTransaction { .. }
+        | Command::Commit { .. }
+        | Command::CancelRequest { .. }
+        | Command::CopyRows { .. }
+        | Command::GetSystemVars { .. }
+        | Command::SetSystemVars { .. }
+        | Command::Terminate { .. } => Ok(()),
+    }
+}
+
+pub fn check_plan(
+    catalog: &impl SessionCatalog,
+    session: &Session,
+    plan: &Plan,
+) -> Result<(), AdapterError> {
+    if !catalog.rbac_checks_enabled() {
+        return Ok(());
+    }
+
+    let role_id = session.role_id();
+    let Some(role) = catalog.try_get_role(role_id) else {
+        // PostgreSQL allows users that have their role dropped to perform some actions,
+        // such as `SET ROLE` and certain `SELECT` queries. We haven't implemented
+        // `SET ROLE` and feel it's safer to force to user to re-authenticate if their
+        // role is dropped.
+        return Err(AdapterError::ConcurrentRoleDrop(role_id.clone()));
+    };
+
+    if session.is_superuser() {
+        return Ok(());
+    }
+
+    if let Some(required_attribute) = generate_plan_attribute(plan) {
+        if !required_attribute.check_role(role) {
+            return Err(AdapterError::Unauthorized(UnauthorizedError::attribute(
+                plan.name().to_string(),
+                required_attribute,
+            )));
+        }
+    }
+
+    Ok(())
+}
+
+/// Generates the attributes required to execute a given plan.
+fn generate_plan_attribute(plan: &Plan) -> Option<Attribute> {
+    match plan {
+        Plan::CreateDatabase(_) => Some(Attribute::CreateDB),
+        Plan::CreateRole(_) | Plan::AlterRole(_) | Plan::DropRoles(_) => {
+            Some(Attribute::CreateRole)
+        }
+        Plan::CreateCluster(_) => Some(Attribute::CreateCluster),
+        Plan::CreateSource(_)
+        | Plan::CreateTable(_)
+        | Plan::CreateMaterializedView(_)
+        | Plan::CreateConnection(_)
+        | Plan::CreateSchema(_)
+        | Plan::CreateClusterReplica(_)
+        | Plan::CreateSecret(_)
+        | Plan::CreateSink(_)
+        | Plan::CreateView(_)
+        | Plan::CreateIndex(_)
+        | Plan::CreateType(_)
+        | Plan::DiscardTemp
+        | Plan::DiscardAll
+        | Plan::DropDatabase(_)
```

Comment:
Does this need CreateDB?

Reply:
No, databases can only be dropped by the database owner (or superusers). They differ from roles because roles have no owner.

Reply:
To be specific, the owner does not need `CREATEDB` to drop the database.